akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolMasterActor.scala

```diff
@@ -0,0 +1,140 @@
+   * @param fm the materializer to use for this pool
+   * @return the newly created actor ref
+   */
+  private[this] def startPoolInterfaceActor(gateway: PoolGateway)(implicit fm: Materializer): ActorRef = {
+    val props = Props(new PoolInterfaceActor(gateway)).withDeploy(Deploy.local)
+    val ref = context.system.actorOf(props, PoolInterfaceActor.name.next())
+    poolStatus += gateway -> PoolInterfaceRunning(ref)
+    poolInterfaces += ref -> gateway
+    context.watch(ref)
+  }
+
+  def receive = {
+
+    // Start or restart a pool without sending it a request. This is used to ensure that
+    // freshly created pools will be ready to serve requests immediately.
+    case s @ StartPool(gateway, materializer) ⇒
+      poolStatus.get(gateway) match {
+        case Some(PoolInterfaceRunning(_)) ⇒
+        case Some(PoolInterfaceShuttingDown(shutdownCompletedPromise)) ⇒
+          shutdownCompletedPromise.future.foreach(_ ⇒ self ! s)
+        case None ⇒
+          startPoolInterfaceActor(gateway)(materializer)
+      }
+
+    // Send a request to a pool. If needed, the pool will be started or restarted.
+    case s @ SendRequest(gateway, request, responsePromise, materializer) ⇒
+      poolStatus.get(gateway) match {
+        case Some(PoolInterfaceRunning(ref)) ⇒
+          ref ! PoolRequest(request, responsePromise)
+        case Some(PoolInterfaceShuttingDown(shutdownCompletedPromise)) ⇒
+          // The request will be resent when the pool shutdown is complete (the first
+          // request will recreate the pool).
+          shutdownCompletedPromise.future.foreach(_ ⇒ self ! s)
+        case None ⇒
+          startPoolInterfaceActor(gateway)(materializer) ! PoolRequest(request, responsePromise)
+      }
+
+    // Shutdown a pool and signal its termination.
+    case Shutdown(gateway, shutdownCompletedPromise) ⇒
+      poolStatus.get(gateway).foreach {
+        case PoolInterfaceRunning(ref) ⇒
+          ref ! PoolInterfaceActor.Shutdown
+          poolStatus += gateway -> PoolInterfaceShuttingDown(shutdownCompletedPromise)
+        case PoolInterfaceShuttingDown(formerPromise) ⇒
+          shutdownCompletedPromise.completeWith(formerPromise.future)
+        case _ ⇒
+          shutdownCompletedPromise.success(Done)
+      }
+
+    // Shutdown all known pools and signal their termination.
+    case ShutdownAll(shutdownCompletedPromise) ⇒
+      val allGateways = poolStatus.keys
+      val counter = new AtomicInteger(allGateways.size)
+      for (gateway ← allGateways)
+        gateway.shutdown().foreach(_ ⇒ if (counter.decrementAndGet() == 0) shutdownCompletedPromise.success(Done))
+
+    // When a pool actor terminate, signal its termination and remove it from our maps.
+    case Terminated(ref) ⇒
+      poolInterfaces.get(ref).foreach { gateway ⇒
+        poolStatus.get(gateway) match {
+          case Some(PoolInterfaceShuttingDown(shutdownCompletedPromise)) ⇒
+            shutdownCompletedPromise.success(Done)
+          case _ ⇒
```

Comment:
An interface pool actor terminates without warning: apart from marking this pool as dead (to be able to restart it), there is nothing else we can do.

The two ignored cases are "pool running and abruptly stopped" and "unknown actor terminates". The first one might need an error in the log, the second one can never happen.